### PR TITLE
Extract trace id from kafka header

### PR DESCRIPTION
**Pull request by RMPall**
### What

This pr attempts to extract trace id from kafka header.This trace id is then used in the logs for tracing purpose.  

### How to review

Please check if the changes make sense. 

### Who can review

Anyone except me.

### CODE REVIEW

File: event/consumer.go [hunk 1]
```diff
@@ -47,31 +47,40 @@ func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler
 // processMessage unmarshals the provided kafka message into an event and calls the handler.
 // After the message is handled, it is committed.
 func processMessage(ctx context.Context, message kafka.Message, handler Handler, cfg *config.Config) {
+	traceID := ctx.Value(kafka.TraceIDHeaderKey)
 	// unmarshal - commit on failure (consuming the message again would result in the same error)
 	event, err := unmarshal(message)
 	if err != nil {
-		log.Error(ctx, "failed to unmarshal event", err)
+		log.Error(ctx, "failed to unmarshal event", err, log.Data{
+			"request-id": traceID,
```
Comment (justinpjose) [inline event/consumer.go:55]: COMMENT 1

You don't want to be adding the `request-id` in the log.Data. When the logs are passed to Kibana, it retrieves the traceID from the ctx (https://github.com/ONSdigital/dp-net/blob/main/request/context_key.go#L9) and adds it in the log visible in Kibana.

So by doing this, the trace id will be displayed twice in the Kibana logs

File: event/consumer.go [hunk 2]
```diff
@@ -47,31 +47,40 @@ func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler
...
 	}
 
-	//nolint: revive, staticcheck, gocritic
-	//TODO: This needs to be looked into log library for customised context details
-	ctx = context.WithValue(ctx, "request-id", event.TraceID)
-
-	log.Info(ctx, "event received", log.Data{"event": event})
+	log.Info(ctx, "event received", log.Data{
+		"event":      event,
+		"request-id": traceID,
```
Comment (justinpjose) [inline event/consumer.go:63]: Same as COMMENT 1

File: event/consumer.go [hunk 3]
```diff
@@ -47,31 +47,40 @@ func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler
...
+	})
 
 	// handle - commit on failure (implement error handling to not commit if message needs to be consumed again)
 	err = handler.Handle(ctx, &event, *cfg)
 	if err != nil {
-		log.Error(ctx, "failed to handle event", err)
+		log.Error(ctx, "failed to handle event", err, log.Data{
+			"request-id": traceID,
```
Comment (justinpjose) [inline event/consumer.go:70]: Same as COMMENT 1

File: event/consumer.go [hunk 4]
```diff
@@ -47,31 +47,40 @@ func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler
...
+		})
 		message.Commit()
 		return
 	}
 
-	log.Info(ctx, "event processed - committing message", log.Data{"event": event})
+	log.Info(ctx, "event processed - committing message", log.Data{
+		"event":      event,
+		"request-id": traceID,
```
Comment (justinpjose) [inline event/consumer.go:78]: Same as COMMENT 1

File: event/consumer.go [hunk 5]
```diff
@@ -47,31 +47,40 @@ func Consume(ctx context.Context, messageConsumer kafka.IConsumerGroup, handler
...
+	})
 	message.Commit()
-	log.Info(ctx, "consumed message committed", log.Data{"event": event})
+	log.Info(ctx, "consumed message committed", log.Data{
+		"event":      event,
+		"request-id": traceID})
```
Comment (justinpjose) [inline event/consumer.go:83]: Same as COMMENT 1

File: event/producer.go [hunk 6]
```diff
@@ -24,13 +24,20 @@ type SearchDataImportProducer struct {
 
 // SearchDataImport produce a kafka message for an instance which has been successfully processed.
 func (p SearchDataImportProducer) SearchDataImport(ctx context.Context, event models.SearchDataImport) error {
+	traceID := ctx.Value(kafka.TraceIDHeaderKey)
 	bytes, err := p.Marshaller.Marshal(event)
 	if err != nil {
 		log.Fatal(ctx, "Marshaller.Marshal", err)
-		return fmt.Errorf(fmt.Sprintf("Marshaller.Marshal returned an error: event=%v: %%w", event), err)
+		return fmt.Errorf(fmt.Sprintf("Marshaller.Marshal returned an error: event=%v: %%w", event), err, log.Data{
+			"request-id": traceID,
```
Comment (justinpjose) [inline event/producer.go:32]: Same as COMMENT 1

File: event/producer.go [hunk 7]
```diff
@@ -24,13 +24,20 @@ type SearchDataImportProducer struct {
...
+		})
 	}
 
 	p.Producer.Channels().Output <- bytes
-	log.Info(ctx, "completed successfully", log.Data{"event": event, "package": "event.SearchDataImport"})
+	log.Info(ctx, "completed successfully", log.Data{
+		"event":      event,
+		"package":    "event.SearchDataImport",
+		"request-id": traceID,
```
Comment (justinpjose) [inline event/producer.go:40]: Same as COMMENT 1

File: event/producer_test.go [hunk 8]
```diff
@@ -145,7 +146,9 @@ func TestProducer_SearchDataImport_MarshalErr(t *testing.T) {
 			err := searchDataImportProducer.SearchDataImport(ctx, expectedSearchDataImportEvent)
 
 			Convey("Then the expected error is returned", func() {
-				expectedError := fmt.Errorf(fmt.Sprintf("Marshaller.Marshal returned an error: event=%v: %%w", expectedSearchDataImportEvent), errors.New("mock error"))
+				expectedError := fmt.Errorf(fmt.Sprintf("Marshaller.Marshal returned an error: event=%v: %%w", expectedSearchDataImportEvent), errors.New("mock error"), log.Data{
+					"request-id": "test",
```
Comment (justinpjose) [inline event/producer_test.go:150]: Same as COMMENT 1

File: handler/handler.go [hunk 9]
```diff
@@ -30,8 +32,10 @@ type ContentPublishedHandler struct {
 
 // Handle takes a single event.
 func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.ContentPublished, cfg config.Config) error {
+	traceID := ctx.Value(kafka.TraceIDHeaderKey)
 	logData := log.Data{
-		"event": cpEvent,
+		"event":      cpEvent,
+		"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:38]: Same as COMMENT 1

File: handler/handler.go [hunk 10]
```diff
@@ -45,15 +49,19 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
 
 		zebedeeContentPublished, err = h.ZebedeeCli.GetPublishedData(ctx, uri)
 		if err != nil {
-			log.Error(ctx, "failed to retrieve published data from zebedee", err)
+			log.Error(ctx, "failed to retrieve published data from zebedee", err, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:53]: Same as COMMENT 1

File: handler/handler.go [hunk 11]
```diff
@@ -45,15 +49,19 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
...
 			return err
 		}
 
 		// byte slice to Json & unMarshall Json
 		var zebedeeData models.ZebedeeData
 		err = json.Unmarshal(zebedeeContentPublished, &zebedeeData)
 		if err != nil {
-			log.Fatal(ctx, "error while attempting to unmarshal zebedee response into zebedeeData", err)
+			log.Fatal(ctx, "error while attempting to unmarshal zebedee response into zebedeeData", err, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:63]: Same as COMMENT 1

File: handler/handler.go [hunk 12]
```diff
@@ -62,23 +70,27 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
...
 		searchData := models.MapZebedeeDataToSearchDataImport(zebedeeData, cfg.KeywordsLimit)
-		searchData.TraceID = cpEvent.TraceID
 		searchData.JobID = cpEvent.JobID
 		searchData.SearchIndex = getIndexName(cpEvent.SearchIndex)
 
 		// Marshall Avro and sending message
 		if sdImportErr := h.Producer.SearchDataImport(ctx, searchData); sdImportErr != nil {
-			log.Error(ctx, "error while attempting to send SearchDataImport event to producer", sdImportErr)
+			log.Error(ctx, "error while attempting to send SearchDataImport event to producer", sdImportErr, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:84]: Same as COMMENT 1

File: handler/handler.go [hunk 13]
```diff
@@ -62,23 +70,27 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
...
+			})
 			return sdImportErr
 		}
 	} else if cpEvent.DataType == DatasetDataType {
 		datasetID, edition, version, getIDErr := getIDsFromURI(cpEvent.URI)
 		if getIDErr != nil {
-			log.Error(ctx, "error while attempting to get Ids for dataset, edition and version", getIDErr)
+			log.Error(ctx, "error while attempting to get Ids for dataset, edition and version", getIDErr, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:92]: Same as COMMENT 1

File: handler/handler.go [hunk 14]
```diff
@@ -88,12 +100,15 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
 		// Make a call to DatasetAPI
 		datasetMetadataPublished, metadataErr := h.DatasetCli.GetVersionMetadata(ctx, "", cfg.ServiceAuthToken, cpEvent.CollectionID, datasetID, edition, version)
 		if metadataErr != nil {
-			log.Error(ctx, "cannot get dataset published contents version %s from api", metadataErr)
+			log.Error(ctx, "cannot get dataset published contents version %s from api", metadataErr, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:104]: Same as COMMENT 1

File: handler/handler.go [hunk 15]
```diff
@@ -120,21 +135,25 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
...
-		datasetVersionMetadata.TraceID = cpEvent.TraceID
 		datasetVersionMetadata.JobID = cpEvent.JobID
 		datasetVersionMetadata.SearchIndex = getIndexName(cpEvent.SearchIndex)
 		datasetVersionMetadata.DataType = "dataset_landing_page"
 
 		// Marshall Avro and sending message
 		if sdImportErr := h.Producer.SearchDataImport(ctx, datasetVersionMetadata); sdImportErr != nil {
-			log.Fatal(ctx, "error while attempting to send DatasetAPIImport event to producer", sdImportErr)
+			log.Fatal(ctx, "error while attempting to send DatasetAPIImport event to producer", sdImportErr, log.Data{
+				"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:149]: Same as COMMENT 1

File: handler/handler.go [hunk 16]
```diff
@@ -120,21 +135,25 @@ func (h *ContentPublishedHandler) Handle(ctx context.Context, cpEvent *models.Co
...
+			})
 			return sdImportErr
 		}
 	} else {
-		log.Info(ctx, "Invalid content data type received, no action")
+		log.Info(ctx, "Invalid content data type received, no action", log.Data{
+			"request-id": traceID,
```
Comment (justinpjose) [inline handler/handler.go:155]: Same as COMMENT 1

**Review by justinpjose — COMMENTED** [review]
Just one tiny comment but in multiple places

File: event/consumer.go (hunk 1 above)
Comment (justinpjose) [reply event/consumer.go:55]: Discussed and this implementation is fine

**Review by justinpjose — APPROVED** [review]

File: event/consumer.go (hunk 2 above)
Comment (justinpjose) [reply event/consumer.go:63]: Discussed and this implementation is fine

File: event/consumer.go (hunk 3 above)
Comment (justinpjose) [reply event/consumer.go:70]: Discussed and this implementation is fine

File: event/consumer.go (hunk 4 above)
Comment (justinpjose) [reply event/consumer.go:78]: Discussed and this implementation is fine

File: event/consumer.go (hunk 5 above)
Comment (justinpjose) [reply event/consumer.go:83]: Discussed and this implementation is fine

File: event/producer.go (hunk 6 above)
Comment (justinpjose) [reply event/producer.go:32]: Discussed and this implementation is fine

File: event/producer.go (hunk 7 above)
Comment (justinpjose) [reply event/producer.go:40]: Discussed and this implementation is fine

File: event/producer_test.go (hunk 8 above)
Comment (justinpjose) [reply event/producer_test.go:150]: Discussed and this implementation is fine

File: handler/handler.go (hunk 9 above)
Comment (justinpjose) [reply handler/handler.go:38]: Discussed and this implementation is fine

File: handler/handler.go (hunk 10 above)
Comment (justinpjose) [reply handler/handler.go:53]: Discussed and this implementation is fine

File: handler/handler.go (hunk 11 above)
Comment (justinpjose) [reply handler/handler.go:63]: Discussed and this implementation is fine

File: handler/handler.go (hunk 12 above)
Comment (justinpjose) [reply handler/handler.go:84]: Discussed and this implementation is fine

File: handler/handler.go (hunk 13 above)
Comment (justinpjose) [reply handler/handler.go:92]: Discussed and this implementation is fine

File: handler/handler.go (hunk 14 above)
Comment (justinpjose) [reply handler/handler.go:104]: Discussed and this implementation is fine

File: handler/handler.go (hunk 15 above)
Comment (justinpjose) [reply handler/handler.go:149]: Discussed and this implementation is fine

**Review by justinpjose — APPROVED** [review]

**Review by justinpjose — DISMISSED** [review]

**Comment by nshumoogum** [comment]
Old PR - closing

**Pull request closed without merging**